bin/oneoff/backfill_data/backfill_script_ids_channel_tokens.rb

```diff
@@ -0,0 +1,199 @@
+    Integer,
+    'Id of last entry to backfill (inclusive).'
+  ) do |end_id|
+    options[:end_id] = end_id
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+puts "Options: #{options}"
+options.freeze
+
+$start_id = options[:start_id]
+$end_id = options[:end_id]
+$is_dry_run = options[:dry_run]
+
+$backfill_count = 0
+$unable_to_backfill = 0
+
+$level_to_script_ids = {}
+
+def update_script_ids
+  puts "Backfilling channel token script_ids..."
+  puts "Script started at #{Time.now}"
+  # find_each uses find_in_batches with a batch size of 1000 (https://apidock.com/rails/ActiveRecord/Batches/find_each)
+  ChannelToken.where(id: $start_id..$end_id).find_each do |channel_token|
+    next if channel_token.script_id.present?
+
+    level = channel_token.level
+    user_id = user_id_for_storage_id(channel_token.storage_id)
+
+    # get all the possible scripts the channel_token could be associated with
+    associated_script_ids = get_associated_script_ids(level)
+
+    # if the level is associated with only one script_level, use that script
+    if associated_script_ids.length == 1
+      script_id = associated_script_ids[0]
+      update_channel_token(channel_token, script_id)
+      next
+    end
+
+    # it's possible the channel token was generated for a logged-out user, in which case no user_id will exist
+    if user_id.present?
+      # get user_scripts for the associated_script_ids, if there is just one user_script, backfill with the
+      # associated script_id
+      script_ids_from_user_scripts = associated_script_ids_by_user_scripts(user_id, associated_script_ids)
+      if script_ids_from_user_scripts.count == 1
+        update_channel_token(channel_token, script_ids_from_user_scripts[0])
+        next
+      end
+
+      # get user_levels for the associated_script_ids, if there is just one user_level, backfill with the
+      # associated script_id
+      script_ids_from_user_levels = associated_script_ids_from_user_levels(user_id, level)
+      if script_ids_from_user_levels.count == 1
+        update_channel_token(channel_token, script_ids_from_user_levels[0])
+        next
+      end
+
+      # if the user has no associated user_levels or user_scripts, it's possible that the user visited a channel backed
+      # level which generated a channel_token and then the user left the page without making any progress. In this case it doesn't
+      # matter which script_id we pick, so backfill with the first associated_script_ids
+      if associated_script_ids.count > 0 && script_ids_from_user_scripts.blank? && script_ids_from_user_levels.blank?
+        update_channel_token(channel_token, associated_script_ids[0])
+        next
+      end
+    end
+
+    record_failed_script_id_backfill(level, channel_token, user_id)
```

Comment:
Should we rescue StandardError at the end of the loop? If any of the logic for backfilling a single ChannelToken fails, will this script terminate or will it log the record that failed and the reason and then move on to the next one?

Reply:
Currently if the logic fails to find a script ID for the channel token, we print that information to the console and move on to the next record. I don't have any explicit error handling, I see your recommendation to check that `update_attributes` succeeded, I will add that.

An open question I had is, what is the best way to log/ keep track of the records that the backfill failed for? I'm assuming that printing to the console isn't the best way

Reply:
I see James used `CDO.log.info`, where does that log to?